File: latest/week3/workshop8_m7/W08_M06_demo.R
# ============================================================
# Workshop 8 — Module 6: Dashboard Design Principles
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

# ── LOAD DATA ────────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")))

# ── GLOBAL PALETTE (Rule 4: Consistent Colour) ──────────
PALETTE <- c(Movie = "#1565C0", "TV Show" = "#E53935")

theme_dash <- theme_minimal(base_size = 9) +
  theme(panel.grid.major = element_line(color = "#F5F5F5"),
    panel.grid.minor = element_blank(),
    plot.title = element_text(face = "bold", size = 10),
    legend.position = "bottom",
    legend.text = element_text(size = 7))
theme_set(theme_dash)

# ══════════════════════════════════════════════════════════
# MOCKUP: Netflix Dashboard Layout (static version)
# Demonstrates the layout grid: KPI → Primary → Secondary → Detail
# ══════════════════════════════════════════════════════════

# ── KPI ROW (Rule 2) ────────────────────────────────────
total <- nrow(nf)
movies_n <- sum(nf$type == "Movie", na.rm = TRUE)
tvshows_n <- sum(nf$type == "TV Show", na.rm = TRUE)
countries_n <- n_distinct(nf$primary_country, na.rm = TRUE)

# KPI cards as mini-plots
make_kpi <- function(label, value, delta, color) {
  ggplot() +
    annotate("text", x = 0.5, y = 0.7,
      label = format(value, big.mark = ","),
      size = 8, fontface = "bold", color = color) +
    annotate("text", x = 0.5, y = 0.35,
      label = paste0(ifelse(delta >= 0, "+", ""), delta, "%"),
      size = 4, fontface = "bold",
      color = ifelse(delta >= 0, "#2E7D32", "#C62828")) +
    annotate("text", x = 0.5, y = 0.12, label = label,
      size = 3, color = "#888") +
    xlim(0, 1) + ylim(0, 1) + theme_void() +
    theme(panel.border = element_rect(color = "#EEEEEE", fill = NA))
}

kpi1 <- make_kpi("Total Titles", total, 12.3, "#333")
kpi2 <- make_kpi("Movies", movies_n, -5.2, "#1565C0")
kpi3 <- make_kpi("TV Shows", tvshows_n, 18.7, "#E53935")
kpi4 <- make_kpi("Countries", countries_n, 3.1, "#7B1FA2")

kpi_row <- kpi1 | kpi2 | kpi3 | kpi4

# ── PRIMARY CHART (Rule 3: largest, most important) ──────
yearly <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)

p_primary <- ggplot(yearly, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1.2) + geom_point(size = 2) +
  scale_color_manual(values = PALETTE) +
  labs(title = "Yearly Additions by Type",
       x = "Year", y = "Titles", color = NULL)

# ── SECONDARY CHART ──────────────────────────────────────
top10 <- nf |> filter(!is.na(primary_country)) |>
  count(primary_country, sort = TRUE) |> slice_max(n, n = 10) |>
  mutate(primary_country = fct_reorder(primary_country, n))

p_secondary <- ggplot(top10, aes(x = primary_country, y = n)) +
  geom_col(fill = "#1565C0", alpha = 0.7, width = 0.6) +
  coord_flip() +
  labs(title = "Top 10 Countries", x = NULL, y = "Titles")

# ── DETAIL ROW (3 small panels) ──────────────────────────
# Detail 1: Type split
type_pct <- nf |> count(type) |> mutate(pct = round(n / sum(n) * 100))
p_detail1 <- ggplot(type_pct, aes(x = type, y = pct, fill = type)) +
  geom_col(width = 0.5, show.legend = FALSE) +
  geom_text(aes(label = paste0(pct, "%")), vjust = -0.5, size = 3, fontface = "bold") +
  scale_fill_manual(values = PALETTE) +
  labs(title = "Type Split", x = NULL, y = "%")

# Detail 2: Rating distribution
p_detail2 <- nf |> filter(!is.na(rating)) |>
  count(rating, sort = TRUE) |> slice_max(n, n = 6) |>
  mutate(rating = fct_reorder(rating, n)) |>
  ggplot(aes(x = rating, y = n)) +
  geom_col(fill = "#2E7D32", alpha = 0.7, width = 0.5) +
  coord_flip() +
  labs(title = "Top Ratings", x = NULL, y = "Count")

# Detail 3: Monthly sparkline
monthly <- nf |>
  filter(!is.na(date_added), year(date_added) >= 2018) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month)
p_detail3 <- ggplot(monthly, aes(x = month, y = n)) +
  geom_line(color = "#1565C0", linewidth = 0.5) +
  geom_area(fill = "#1565C0", alpha = 0.05) +
  scale_x_date(date_labels = "%Y") +
  labs(title = "Monthly Trend (2018–2021)", x = NULL, y = "Titles")

# ── COMPOSE FULL DASHBOARD ───────────────────────────────
dashboard <- kpi_row /
  (p_primary | p_secondary) /
  (p_detail1 | p_detail2 | p_detail3) +
  plot_annotation(
    title = "Netflix Content Dashboard",
    subtitle = "Layout: KPI row → Primary + Secondary → Detail panels (Few's grid)",
    tag_levels = list(c("", "", "", "",
      "Primary", "Secondary", "Detail 1", "Detail 2", "Detail 3")))

ggsave("output/dashboard_mockup.png", dashboard, width = 14, height = 12, dpi = 300)

# ── ANTI-PATTERN: BAD DASHBOARD ──────────────────────────
# Deliberately violate all 6 rules
p_bad1 <- ggplot(yearly, aes(x = factor(year_added), y = n, fill = type)) +
  geom_col(position = "dodge", color = "black") +
  scale_fill_manual(values = c(Movie = "#FF6347", "TV Show" = "#32CD32")) + # wrong colours!
  theme_gray() + labs(title = "ANTI-PATTERN 1: Wrong colours + borders")

p_bad2 <- ggplot(yearly, aes(x = year_added, y = n, color = type)) +
  geom_line(linewidth = 1) +
  scale_color_manual(values = c(Movie = "#4169E1", "TV Show" = "#FFD700")) + # inconsistent!
  theme_gray() + labs(title = "ANTI-PATTERN 2: Different palette same data")

ggsave("output/anti_patterns.png",
  (p_bad1 | p_bad2) +
    plot_annotation(title = "Anti-Patterns: Colour Chaos (same data, different palettes per chart)"),
  width = 14, height = 5, dpi = 300)

# ── FEW'S RULES CHECKLIST ────────────────────────────────
cat("\n── Few's Dashboard Design Rules ──\n")
rules <- tibble(
  rule = 1:6,
  name = c("One screen", "KPIs at top", "5–7 charts max",
    "Consistent colour", "Filter sidebar", "Descriptive titles"),
  check = c(
    "No scrolling required?",
    "Big numbers visible in first 2 seconds?",
    "Count of charts ≤ 7?",
    "Same colour = same meaning across all panels?",
    "All inputs in sidebar or top bar (not inline)?",
    "Titles describe the metric (not the finding)?"))
print(rules, n = 6)

cat("\n── All W08-M06 R outputs saved ──\n")

File: latest/week3/workshop8_m6/W08_M06_demo.py
"""W08-M06: Dashboard Design Principles — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
from matplotlib.gridspec import GridSpec
os.makedirs("output", exist_ok=True)

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()

# GLOBAL PALETTE (Rule 4)
PALETTE = {"Movie": "#1565C0", "TV Show": "#E53935"}

yearly = nf.query("2015<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")

# ══════════════════════════════════════════════════════════
# DASHBOARD MOCKUP (static matplotlib version)
# Demonstrates the layout grid
# ══════════════════════════════════════════════════════════
fig = plt.figure(figsize=(14, 12))
gs = GridSpec(4, 4, figure=fig, hspace=0.40, wspace=0.35,
    height_ratios=[0.8, 2.5, 2.5, 2])

# ── KPI ROW ──────────────────────────────────────────────
kpis = [
    ("Total Titles", f"{len(nf):,}", "+12.3%", "#333"),
    ("Movies", f"{(nf.type=='Movie').sum():,}", "–5.2%", "#1565C0"),
    ("TV Shows", f"{(nf.type=='TV Show').sum():,}", "+18.7%", "#E53935"),
    ("Countries", f"{nf['primary_country'].nunique()}", "+3.1%", "#7B1FA2"),
]
for i, (label, value, delta, color) in enumerate(kpis):
    ax = fig.add_subplot(gs[0, i])
    ax.text(0.5, 0.65, value, ha="center", va="center",
        fontsize=18, fontweight="bold", color=color, transform=ax.transAxes)
    delta_color = "#2E7D32" if delta.startswith("+") else "#C62828"
    ax.text(0.5, 0.30, delta, ha="center", va="center",
        fontsize=10, fontweight="bold", color=delta_color, transform=ax.transAxes)
    ax.text(0.5, 0.05, label, ha="center", va="center",
        fontsize=8, color="#888", transform=ax.transAxes)
    ax.axis("off")
    ax.add_patch(plt.Rectangle((0.02, 0.02), 0.96, 0.96, fill=False,
        edgecolor="#EEEEEE", lw=1, transform=ax.transAxes))

# ── PRIMARY CHART ────────────────────────────────────────
ax = fig.add_subplot(gs[1, :2])
for t, c in PALETTE.items():
    sub = yearly.query("type==@t")
    ax.plot(sub.year_added, sub.n, "-o", color=c, lw=1.5, markersize=4, label=t)
ax.legend(fontsize=7); ax.set_title("Yearly Additions by Type", fontsize=10, fontweight="bold")
ax.set_ylabel("Titles"); ax.grid(axis="y", alpha=0.2)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

# ── SECONDARY CHART ──────────────────────────────────────
ax = fig.add_subplot(gs[1, 2:])
top10 = nf.dropna(subset=["primary_country"])["primary_country"].value_counts().head(10).sort_values()
ax.barh(top10.index, top10.values, color="#1565C0", alpha=0.7, height=0.6)
ax.set_title("Top 10 Countries", fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

# ── DETAIL ROW ───────────────────────────────────────────
# Detail 1: Type split
ax = fig.add_subplot(gs[2, :2])
type_split = nf["type"].value_counts()
ax.bar(type_split.index, type_split.values/type_split.sum()*100,
    color=[PALETTE.get(t, "#888") for t in type_split.index], width=0.4)
for i, (t, v) in enumerate(zip(type_split.index, type_split.values/type_split.sum()*100)):
    ax.text(i, v+1, f"{v:.0f}%", ha="center", fontsize=9, fontweight="bold")
ax.set_title("Type Split (%)", fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

# Detail 2: Top ratings
ax = fig.add_subplot(gs[2, 2:])
top_ratings = nf.dropna(subset=["rating"])["rating"].value_counts().head(6).sort_values()
ax.barh(top_ratings.index, top_ratings.values, color="#2E7D32", alpha=0.7, height=0.5)
ax.set_title("Top Ratings", fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

# Detail 3: Monthly sparkline
ax = fig.add_subplot(gs[3, :])
monthly = (nf.dropna(subset=["date_added"]).query("date_added.dt.year>=2018")
    .set_index("date_added").resample("ME").size())
ax.plot(monthly.index, monthly.values, color="#1565C0", lw=0.8)
ax.fill_between(monthly.index, monthly.values, alpha=0.05, color="#1565C0")
from matplotlib.dates import DateFormatter
ax.xaxis.set_major_formatter(DateFormatter("%Y-%m"))
ax.set_title("Monthly Trend (2018–2021)", fontsize=10, fontweight="bold")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)

fig.suptitle("Netflix Content Dashboard (Static Mockup)\n"
    "Layout: KPI row → Primary + Secondary → Detail panels",
    fontsize=14, fontweight="bold")
plt.savefig("output/dashboard_mockup.png", dpi=300, bbox_inches="tight"); plt.close()

# ══════════════════════════════════════════════════════════
# DASH KPI CARD TEMPLATE
# ══════════════════════════════════════════════════════════
kpi_template = '''
import dash_bootstrap_components as dbc
from dash import html

def make_kpi(title, value, delta):
    """Reusable KPI card component."""
    color = "success" if delta >= 0 else "danger"
    arrow = "↑" if delta >= 0 else "↓"
    return dbc.Card(dbc.CardBody([
        html.H6(title, className="text-muted mb-1",
            style={"fontSize": "0.8rem"}),
        html.H3(f"{value:,}", className="fw-bold mb-1"),
        html.P(f"{arrow} {abs(delta):.1f}%",
            className=f"text-{color} mb-0",
            style={"fontSize": "0.9rem"}),
    ]), className="text-center shadow-sm")

# Usage in layout:
kpi_row = dbc.Row([
    dbc.Col(make_kpi("Total Titles", 8807, 12.3), width=3),
    dbc.Col(make_kpi("Movies", 6131, -5.2), width=3),
    dbc.Col(make_kpi("TV Shows", 2676, 18.7), width=3),
    dbc.Col(make_kpi("Countries", 78, 3.1), width=3),
], className="mb-3")
'''
with open("output/kpi_template.py", "w") as f:
    f.write(kpi_template)

# ══════════════════════════════════════════════════════════
# FEW'S RULES + ANTI-PATTERNS
# ══════════════════════════════════════════════════════════
print("=== Few's Six Dashboard Design Rules ===")
rules = [
    ("1. One screen", "No scrolling. Everything visible at once."),
    ("2. KPIs at top", "Big numbers first. Charts below for context."),
    ("3. 5–7 charts max", "Cognitive limit. More = overwhelm."),
    ("4. Consistent colour", "Same colour = same meaning across all panels."),
    ("5. Filter sidebar", "Inputs on left or top. Never inline with charts."),
    ("6. Descriptive titles", "Dashboard = reader-driven. State the metric, not the finding."),
]
for name, desc in rules:
    print(f"  {name}: {desc}")

print("\n=== Five Anti-Patterns ===")
antis = [
    ("Chartjunk dashboard", "→ Flat, minimal, data-ink focus"),
    ("Scroll-of-death", "→ Max 5–7 charts; use tabs"),
    ("Colour chaos", "→ Global palette applied everywhere"),
    ("No KPIs", "→ Add KPI row at top"),
    ("Filter everywhere", "→ Sidebar or top filter bar"),
]
for name, fix in antis:
    print(f"  {name} {fix}")

print("\nAll W08-M06 Python outputs saved")
